_source/_posts/2021-08-09-reactive-java.md

```diff
@@ -0,0 +1,808 @@
+  artifactId==reactive-service \
+  name==reactive-service \
+  packageName==com.okta.developer.reactive \
+  javaVersion==11 \
+  dependencies==webflux,okta
+```
+Unzip the project:
+
+```shell
+unzip reactive-service.zip
+cd reactive-service
+```
+
+{% include setup/cli.md type="web" framework="Okta Spring Boot Starter" %}
+
+
+Create the package `com.okta.developer.reactive.service`, and add the `SecureRandomService` interface:
+
+```java
+package com.okta.developer.reactive.service;
+
+import reactor.core.publisher.Mono;
+
+public interface SecureRandomService {
+
+    Mono<Integer> getRandomInt();
+}
+```
+
+And the implementation `SecureRandomServiceImpl`:
+
+```java
+package com.okta.developer.reactive.service;
+
+import org.springframework.stereotype.Service;
+import reactor.core.publisher.Mono;
+
+import java.security.SecureRandom;
+
+@Service
+public class SecureRandomServiceImpl implements SecureRandomService {
+
+    private SecureRandom secureRandom;
+
+    public SecureRandomServiceImpl() {
+        secureRandom = new SecureRandom();
+    }
+
+    @Override
+    public Mono<Integer> getRandomInt() {
+        return Mono.just(secureRandom.nextInt());
+    }
+}
+```
+
+Think for a moment about the `getRandomInt` implementation. It is hard to resist the temptation of just wrapping whatever in a publisher. `Mono.just(T data)` creates a `Mono` that will emit the specified item. But the item is **captured at instantiation time**. This means the blocking code will be invoked on assembly time, and it might be in the context of an event loop thread.
+
+For now, let's move forward and create the package `com.okta.developer.reactive.controller`, and add a `SecureRandom` class for the data:
+
+```java
+package com.okta.developer.reactive.controller;
+
+public class SecureRandom {
+
+    private String value;
+
+    public String getValue() {
+        return value;
+    }
+
+    public void setValue(String value) {
+        this.value = value;
+    }
+
+    public SecureRandom(Integer value){
+        this.value = value.toString();
+    }
+}
+```
+Create the `SecureRandomController` class:
+
+```java
+package com.okta.developer.reactive.controller;
+
+import com.okta.developer.reactive.service.SecureRandomService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RestController;
+import reactor.core.publisher.Mono;
+
+@RestController
+public class SecureRandomController {
+
+    @Autowired
+    private SecureRandomService secureRandomService;
+
+
+    @GetMapping("/random")
+    public Mono<SecureRandom> getSecureRandom(){
+        return secureRandomService.getRandomInt().map(i -> new SecureRandom(i));
+    }
+}
+```
+
+Add `spring-security-test` dependency to the pom:
+
+```xml
+<dependency>
+    <groupId>org.springframework.security</groupId>
+    <artifactId>spring-security-test</artifactId>
+    <version>5.4.5</version>
```

Comment:
```suggestion
```

This will get pulled in from the spring parent pom / spring-boot-dependencies import